lib/Optimizer/Dialect/CC/CCOps.cpp

```diff
@@ -214,6 +214,12 @@ OpFoldResult cudaq::cc::CastOp::fold(FoldAdaptor adaptor) {
       auto val = attr.getInt();
       if (isa<IntegerType>(ty)) {
         auto width = ty.getIntOrFloatBitWidth();
+        auto srcTy = getValue().getType();
```

Comment:
The bug here is that this is not respecting the `signed` and `unsigned` flag on the cast. This is the correct code for truncation only.

What needs to be added here are the cases as below with `getSint()` and `getZint()` selectors. I suspect the `getSint()` case works with the original code as well. For `getZInt()`, we should simply mask off higher order bits. That is,
```c++
if (fromWidth < 64)
  val &= (1 << fromWidth) - 1;
```

Reply:
I used c++ to cast the const op value to the original type, to avoid possible bugs... let me know if you prefer using bit masking instead.

Reply:
see the last comment.